### PR TITLE
Fix inconsistent return types of `quantile` etc.

### LINKED ISSUE #1798: Inconsistent behaviour of `truncate` for discrete distributions
When truncating a `Geometric` distributions, `rand` will return different types depending on the bounds chosen. For example, values sampled from `truncated(Geometric(0.6); lower = 2)` are floats, while values sampled from `truncated(Geometric(0.6); upper = 4)` are integers.
I guess this is due to https://github.com/JuliaStats/Distributions.jl/blob/7e232ca86c56c4b2e1e7a44b16aa23c26668af3d/src/truncate.jl#L211-L229 using the float-returning `quantile` / `invlogcdf` in special cases.

### PR DESCRIPTION
**Pull request by devmotion**
Currently, the return type of `quantile` etc. is inconsistent, in particular for discrete distributions. For many distributions (and in particular those for which we fall back to StatsFuns/Rmath) the return type is explicitly to converted to `Int` to match the type of the variates. However, the behaviour of some distributions such as `Geometric` or `Bernoulli` is inconsistent: The return type of `quantile` does not match the type of the variates.

This PR fixes the inconsistency. Moreover, I added tests to avoid regressions and made an existing test a bit stricter. These tests revealed a few additional problems (e.g. caused by the Rmath behaviour that `StatsFuns.RFunctions.binominvcdf(10, 1.0, 0.0) = 0` and `StatsFuns.RFunctions.hyperinvlogccdf(2, 2, 2, -Inf) = NaN` which so far we have hidden by skipping them in the tests).

Fixes #1798.

### CODE REVIEW

File: src/univariates.jl [hunk 1]
```diff
@@ -619,6 +628,38 @@ function integerunitrange_logccdf(d::DiscreteUnivariateDistribution, x::Integer)
...
 
+### Error messages
+
+const QUANTILE_ARG_ERROR = "`Distributions.quantile(d, p)` only accepts arguments `p` in `[0, 1]`." 
+const CQUANTILE_ARG_ERROR = "`Distributions.cquantile(d, p)` only accepts arguments `p` in `[0, 1]`." 
+const INVLOGCDF_ARG_ERROR = "`Distributions.invlogcdf(d, lp)` only accepts non-negative arguments `lp`." 
+const INVLOGCCDF_ARG_ERROR = "`Distributions.invlogccdf(d, lp)` only accepts non-negative arguments `lp`." 
+
+function _check_quantile_arg(p::Real)
+    if !(zero(p) <= p <= oneunit(p))
```
Comment (nalimilan) [inline src/univariates.jl:639]: Does `oneunit` really make sense here? For `Real`s AFAIK it's the same as `one`. (I also suspect that for the compiler it's equivalent to `0 <= p <= 1` but I guess it's safer to use `zero` and `one`.)

File: src/univariates.jl [hunk 2]
```diff
@@ -619,6 +628,38 @@ function integerunitrange_logccdf(d::DiscreteUnivariateDistribution, x::Integer)
     return result
 end
 
+### Error messages
+
+const QUANTILE_ARG_ERROR = "`Distributions.quantile(d, p)` only accepts arguments `p` in `[0, 1]`."
```
Comment (nalimilan) [inline src/univariates.jl:633]: While you're at it, how about adding `(got $p)`? That's generally very useful for users.

Also, I just noticed that `?quantile` calls this argument `q` so it would be good to use the same name (maybe `p` for consistency with Statistics?).

File: src/univariates.jl [hunk 3]
```diff
@@ -619,6 +628,38 @@ function integerunitrange_logccdf(d::DiscreteUnivariateDistribution, x::Integer)
     return result
 end
 
+### Error messages
+
+const QUANTILE_ARG_ERROR = "`Distributions.quantile(d, p)` only accepts arguments `p` in `[0, 1]`." 
+const CQUANTILE_ARG_ERROR = "`Distributions.cquantile(d, p)` only accepts arguments `p` in `[0, 1]`." 
+const INVLOGCDF_ARG_ERROR = "`Distributions.invlogcdf(d, lp)` only accepts non-negative arguments `lp`." 
+const INVLOGCCDF_ARG_ERROR = "`Distributions.invlogccdf(d, lp)` only accepts non-negative arguments `lp`."
```
Comment (nalimilan) [inline src/univariates.jl:636]: "Non-negative" doesn't seem to reflect the check `!(lp <= zero(lp))`. Isn't that "strictly positive" instead?

BTW, for some distributions (e.g. `Geometric`) it was allowed to pass zero and now it throws an error. Is that intended?

File: src/univariate/discrete/negativebinomial.jl [hunk 4]
```diff
@@ -112,10 +112,30 @@ cdf(d::NegativeBinomial, x::Real) = nbinomcdf(d.r, d.p, x)
...
+    _check_quantile_arg(q)
+    return nbinominvcdf(d.r, d.p, q)
+end
+function cquantile(d::NegativeBinomial, q::Real)::Int
+    _check_cquantile_arg(q)
+    return nbinominvccdf(d.r, d.p, q)
+end
+function invlogcdf(d::NegativeBinomial, lq::Real)::Int
+    _check_invlogcdf_arg(lq)
+    if isinf(lq)
```
Comment (nalimilan) [inline src/univariate/discrete/negativebinomial.jl:125]:
```suggestion
    if isinf(lq) # -Inf
```
(I didn't get it at first.)

File: src/univariate/discrete/negativebinomial.jl [hunk 5]
```diff
@@ -112,10 +112,30 @@ cdf(d::NegativeBinomial, x::Real) = nbinomcdf(d.r, d.p, x)
...
+        return nbinominvcdf(d.r, d.p, zero(lq))
+    else
+        return nbinominvlogcdf(d.r, d.p, lq)
+    end
+end
+function invlogccdf(d::NegativeBinomial, lq::Real)::Int
+    _check_invlogccdf_arg(lq)
+    if isinf(lq)
```
Comment (nalimilan) [inline src/univariate/discrete/negativebinomial.jl:133]:
```suggestion
    if isinf(lq) # -Inf
```

File: src/univariate/discrete/noncentralhypergeometric.jl [hunk 6]
```diff
@@ -9,23 +9,23 @@ abstract type NoncentralHypergeometric{T<:Real} <: DiscreteUnivariateDistributio
...
+        range = reverse(range)
+    end
+
+    # Support must not be empty
+    r, state = iterate(range)
+    qsum = pdf(d, r)
+    while qsum < q && (r_state = iterate(range, state)) !== nothing
+        r, state = r_state
+        qsum += pdf(d, r)
     end
```
Comment (nalimilan) [inline src/univariate/discrete/noncentralhypergeometric.jl:27]: Wouldn't it be simpler to keep the current approach and just check that `!isempty(range)`? I don't find the `iterate` approach super readable.

File: src/univariates.jl [hunk 7]
```diff
@@ -396,7 +396,10 @@ quantile(d::UnivariateDistribution, p::Real)
 
 The complementary quantile value, i.e. `quantile(d, 1-q)`.
 """
-cquantile(d::UnivariateDistribution, p::Real) = quantile(d, 1.0 - p)
+function cquantile(d::UnivariateDistribution, p::Real)
+    _check_cquantile_arg(p)
+    quantile(d, 1 - p)
```
Comment (nalimilan) [inline src/univariates.jl:401]: Or even:
```suggestion
    quantile(d, one(p) - p)
```

File: src/univariates.jl [hunk 8]
```diff
@@ -649,10 +690,34 @@ macro _delegate_statsfuns(D, fpre, psyms...)
...
+            _check_cquantile_arg(q)
+            return $(finvccdf)($(pargs...), q)
+        end
+
+        # There is a bug in some functions in StatsFuns.RFunctions/Rmath:
+        # For `lq = -Inf` they return `NaN`
+        # We work around this issue by falling back to quantile/cquantile
+        function $Distributions.invlogcdf(d::$D, lq::Real)::$T
+            _check_invlogcdf_arg(lq)
+            if isinf(lq)
```
Comment (nalimilan) [inline src/univariates.jl:707]:
```suggestion
            if isinf(lq) # -Inf
```

File: src/univariates.jl [hunk 9]
```diff
@@ -649,10 +690,34 @@ macro _delegate_statsfuns(D, fpre, psyms...)
...
+                return $(finvcdf)($(pargs...), zero(lq))
+            else
+                return $(finvlogcdf)($(pargs...), lq)
+            end
+        end
+        function $Distributions.invlogccdf(d::$D, lq::Real)::$T
+            _check_invlogccdf_arg(lq)
+            if isinf(lq)
```
Comment (nalimilan) [inline src/univariates.jl:715]:
```suggestion
            if isinf(lq) # -Inf
```

File: src/univariate/discrete/negativebinomial.jl [hunk 10]
```diff
@@ -112,10 +112,30 @@ cdf(d::NegativeBinomial, x::Real) = nbinomcdf(d.r, d.p, x)
...
-invlogccdf(d::NegativeBinomial, lq::Real) = convert(Int, nbinominvlogccdf(d.r, d.p, lq))
+function quantile(d::NegativeBinomial, q::Real)::Int
+    _check_quantile_arg(q)
+    return nbinominvcdf(d.r, d.p, q)
+end
+function cquantile(d::NegativeBinomial, q::Real)::Int
+    _check_cquantile_arg(q)
+    return nbinominvccdf(d.r, d.p, q)
+end
+function invlogcdf(d::NegativeBinomial, lq::Real)::Int
```
Comment (nalimilan) [inline src/univariate/discrete/negativebinomial.jl:123]: Aren't this method and the next one redundant with the new general definition from `_delegate_statsfuns`?

File: test/testutils.jl [hunk 11]
```diff
@@ -447,6 +447,7 @@ end
 
 
 function test_evaluation(d::DiscreteUnivariateDistribution, vs::AbstractVector, testquan::Bool=true)
```
Comment (nalimilan) [inline test/testutils.jl:449]: Are there systematic tests somewhere that for all distributions, `p` and `lp` throw when they are not in the expected range? That seems easy to do in a loop for all corner cases (0.0, -0.0, 1, Inf, -Inf, NaN...).

**Review by nalimilan — COMMENTED** [review]
Since it appears you need a review, here's one, though I'm not familiar with the package. ;-)

Just to make sure everybody is on the same page: currently many common distributions (`Normal`, `Geometric`, `Exponential`, `Cauchy`, `Bernoulli`...) support `quantile(d, NaN) === NaN` and similar, while others throw an error. With this PR it's always an error. That's cleaner, but it's breaking so that requires a new major release (time to tag 1.0?).

File: src/univariates.jl (hunk 1 above)
Comment (dkarrasch) [reply src/univariates.jl:639]: For unitful numbers, `one` gives the unitless 1 so that `1*p = p` and `one(p)` and `p` may not even be comparable. Instead, `oneunit` returns a number of the same unit as `p` and should be comparable.

Comment (dkarrasch) [reply src/univariates.jl:639]: I forgot if unitful numbers subtype `Real` or just `Number`, so my comment may be irrelevant...

Comment (simsurace) [reply src/univariates.jl:639]: `Number` AFAIK, not `Real`

Comment (devmotion) [reply src/univariates.jl:639]: That's not enforced anywhere, so AFAICT in principle also unitful numbers could subtype `Real`. Unitfuls.jl, however, does only subtype `Number`, not `Real`.

Comment (dkarrasch) [reply src/univariates.jl:639]: I still think that - regardless whether it makes a computational difference - `oneunit` is conceptually the right thing to use here: `p` and `zero(p)` are of same "unit"/type, `one(p)` may not.

Comment (nalimilan) [reply src/univariates.jl:639]: Not a big deal anyway so it shouldn't hold this PR back.
